DataScience/AzureStorageDownloader.py

```diff
@@ -1,127 +1,176 @@
+        config = configparser.ConfigParser()
+        config.read('ds.config')
+        account_name = config['DecisionService']['AzureBlobStorageAccountName']
+        account_key = config['DecisionService']['AzureBlobStorageAccountKey']
+    
+    if len(account_key.replace('%2b','+')) != 88:
+        print("Invalid Azure Storage Account Key!")
         sys.exit()
+    
+    if version == 1: # using LogDownloader api
+        if not overwrite and os.path.isfile(output_fp):
+            print('File ({}) already exits, not downloading'.format(output_fp))
+        else:
+            print('Destination: {}\nDownloading...'.format(output_fp), end='')
+            if dry_run:
+                print(' dry_run!')
+            else:
+                try:
+                    url = LogDownloaderURL.format(ACCOUNT_NAME=account_name, ACCOUNT_KEY=account_key.replace('+','%2b'), CONTAINER=container, START_DATE=start_date.strftime("%Y-%m-%d"), END_DATE=end_date.strftime("%Y-%m-%d"))
+                    r = requests.post(url)
+                    open(output_fp, 'wb').write(r.content)
+                    print(' Done!')
+                except Exception as e:
+                    print(' Error: {}'.format(e))
+        
+    else: # using BlockBlobService python api
+        bbs = BlockBlobService(account_name=account_name, account_key=account_key.replace('%2b','+'))
+
+        # List all blobs and download them one by one
+        blobs = bbs.list_blobs(container)
+        for blob in blobs:
+            if '/data/' not in blob.name:
+                continue
+            fp = os.path.join(log_dir, container, blob.name.replace('/','_'))
+            if os.path.isfile(fp) and not overwrite:
+                continue
+            
+            blob_day = datetime.datetime.strptime(blob.name.split('/data/', 1)[1].split('_', 1)[0], '%Y/%m/%d')
+            if (start_date and blob_day < start_date) or (end_date and end_date <= blob_day):
+                continue
 
+            try:
+                bp = bbs.get_blob_properties(container, blob.name)
+                curr_time = datetime.datetime.now(datetime.timezone.utc)
+                diff_time = curr_time-bp.properties.last_modified
+                will_download = (not skip_current) or diff_time > datetime.timedelta(0, 600, 0)    # download blob if not modified in the last 10 min
+                if will_download:
+                    print("\nBlob {}\nUncompressed Size: {:.3f} MB\nLast modified: {}\nCurrent time: {}\nDiff from current time: {} - Will download: {}".format(blob.name, bp.properties.content_length/(1024**2), bp.properties.last_modified, curr_time, diff_time, will_download))
+                    print('Destination: {}\nDownloading...'.format(fp), end='')
+                    if dry_run:
+                        print(' dry_run!')
+                    else:
+                        t0 = time.time()
+                        bbs.get_blob_to_path(container, blob.name, fp)
```

Comment:
do you already get intermediate progress? if not this method provides a progress_callback invoked regularly to give user feedback

Reply:
I didn't implement intermediate progress. I will look into it